doc/parent_child_spec.mld

```diff
@@ -157,3 +161,34 @@ and we shall get
 ]}
 
 For more about [odoc] commands, simply invoke [odoc --help] in your shell.
+
+{1 Convention for installed packages}
+
+Locally, the build system can make arbitrary complex documentation page
+hierarchies.
+However, the generated HTML documentation is generally not installed as part of
+a package. Instead the documentation source code made of [.mld] pages is
+installed and might be used by a different driver.
+
+{2 Convention}
+
+In order for drivers to build consistent documentation for a package, the
+following convention should be followed.
+
+- [.mld] pages are installed in a package's [share] directory, under the
+  [odoc-pages] sub-directory.
+- [index.mld] is the parent of every other pages. The driver can freely rename
+  this page, for example, it can be named after the package.
```

Comment:
The renaming bit doesn't work, it prevents from making links from other pages into the `index` page ([example](https://github.com/dbuenzli/cmarkit/blob/187fe9daeb8d4449453cc4c73657dd9f73b54be3/src/cmarkit_html.mli#L11)).

Reply:
The index of package page is not generated using Odoc ? It can't make references to `{!page-index}`, as that would be very ambiguous.

Though I believe we need references through the page hierarchy like `{!page-packages.odoc.index}` but even then, the `index` part isn't wanted.

Reply:
Also, we'll want to allow deeper hierarchies in the future and our first though is to generalize the `index.mld` renaming:

If you have `foo/bar/index.mld`, it makes a page named `foo/bar.html` and have `foo/bar/*.mld` as children.

Reply:
@dbuenzli you're right, linking to the index is currently broken. We can fix this by having links to `page-index` resolve to the parent, but it all feels a little hacky. If anyone has any other bright ideas I'll be very happy to hear suggestions.

>The index of package page is not generated using Odoc ? It can't make references to {!page-index}, as that would be very ambiguous.

It's only ambiguous knowing how the various drivers work - it's entirely unambiguous in the context of having a flat hierarchy of mld pages per package - ie, the view that the package author has. There can be only 1 `index.mld` file, as they're all stored in the same directory (and dune at least enforces this - I presume other packaging tools do similarly.) When we have support for user-written hierarchical docs, we can be more explicit in what to name your mlds, how the hierarchy works, how to make links between pages correctly and whatnot.

Reply:
>I notice for [the example](https://github.com/dbuenzli/cmarkit/blob/187fe9daeb8d4449453cc4c73657dd9f73b54be3/src/cmarkit_html.mli#L11) I mention that while the link gets [correctly rendered](https://erratique.ch/software/cmarkit/doc/Cmarkit_html/index.html) in the brzo driver which I use for docsets on erratique.ch, it's neither rendered by odig nor by [ocaml.org](https://ocaml.org/p/cmarkit/latest/doc/Cmarkit_html/index.html).

I'm slightly surprised that odig doesn't render this link? I know exactly why ocaml.org doesn't!

Reply:
> For cross packages links I don't really care if there's an additional .index needed.

The problem with the `.index` in `{!packages.odoc.index}` is that it implies that `{!packages.odoc}` links to something. I think we want to pretend that `odoc/index.mld` is named `odoc.mld`.

> For cross packages links I don't really care if there's an additional .index needed. However one thing that is important is not to have to hard code your package/parent name everywhere. Even for your example of deeper hierarchies you want to be able to refer to your parent's index page without having to name it.

I think we need a syntax for references to the parent page. We also need to take into account the parent page in case of conflict (eg. if `foo/bar.mld` contains `{!baz}`, `foo/baz.mld` should be considered before `baz.mld`).

But I also think naming the parent is fine. Deciding that `foo/index.mld` is in fact named `foo` is not a problem to me. I would see `{!page-foo}` written from either `baz.mld` or `foo/bar.mld`.

Reply:
We will need a syntax for references to the parent page, but we don't for the current standard in mlds for a package, because there *aren't any parent pages* - the only thing that works today is that there is a flat set of pages per opam package. I think we should focus here on what currently works (or _should_ work!)

Reply:
From the perspective of the package author, it's very weird that if you have a pages `foo.mld` `bar.mld` and `index.mld` that you can link to the first two with the references `{!page-foo}` and `{!page-bar}`, but not to `{!page-index}` - furthermore it is _entirely impossible_ to link to that page because you've no idea what the driver might have named it - in the dune rules I'm working on right now it gets renamed to `<package name>.mld` but in voodoo it's renamed as `doc.mld`. This is clearly a bug, and we should make references to `{!page-index}` work as expected somehow.

Reply:
> I'm slightly surprised that odig doesn't render this link? I know exactly why ocaml.org doesn't!

The logs says (using `odoc` 2.2.1):
```
File "src/cmarkit_latex.mli", line 11, characters 8-44:
Warning: Failed to resolve reference unresolvedroot(index).quick Couldn't find page "quick"
```

> The problem with the .index in {!packages.odoc.index} is that it implies that {!packages.odoc} links to something.

Not really, it can perfectly be an undefined reference.

> But I also think naming the parent is fine. Deciding that `foo/index.mld` is in fact named `foo` is not a problem to me. I would see `{!page-foo}` written from either `baz.mld` or `foo/bar.mld`

In my opinion these kind of shortcuts tend to bring in more problems than they solve because now you need to say something about what happens when you have both `foo.mld` and `foo/index.mld`. 

Also isn't the problem here that drivers do want to decide on the parents name (e.g. for versioning on ocaml.org ?).

Reply:
I remain quite unconvinced about all this `.mld` file renaming business occuring in all these drivers.

Just thinking aloud, didn't fully think this over.

It seems assets references proposed in #1002 have some kind of search procedure (not sure it's a good idea, but why not) what if we start treating pages references like assets ? 

Then if we add the possibility to add empty pages we get:

```
pkg <-- empty page created by the driver
└── index.mld
    ├── Mymodule
    ├── myimage.gif
    └── mypage.mld
```

Now a reference in for `page-index` first looks up in your level, and then up and recursively (that's what is done in #1002 IIUC).

Reply:
(The slighly annoying bit is that package cross references go through `index` then e.g. 
`{!pkg.index.mypage}`)

Reply:
Ah but no in fact we simply do: 
```
pkg <-- empty, non materialized page created by the driver
├── Mymodule
├── index.mld
├── myimage.gif
└── page.mld
```
and simply decrete that `index.mld` is the landing page shown to users. 

Note that this essentially comes back to the old `--pkg` option.

Reply:
> I think we should focus here on what currently works (or should work!)

I think it's useful to discuss more because we'll want deep hierarchy soon and it would be a shame to create conflicts now.

> furthermore it is entirely impossible to link to that page because you've no idea what the driver might have named it

Maybe it's reasonable to say that we never want to reference a parent page by its name as it was said above ?
Though, the convention should tell how it should be renamed.

> In my opinion these kind of shortcuts tend to bring in more problems than they solve because now you need to say something about what happens when you have both foo.mld and foo/index.mld.

Defining what happens in this case does not seem worse than having undefined references exist.

I think good looking URLs, then good looking references come before work in the driver.

> It seems assets references proposed in #1002 have some kind of search procedure (not sure it's a good idea, but why not) what if we start treating pages references like assets ?

Yes I think we want that! It's currently impossible to solve a page name conflict without limiting what can be linked to (by limiting the search path, as it's done currently).
It's also a reason why we need dot references, a page outside of this hierarchy could do `{!page-pkg.index.mypage}`.

We also want to do:

```
pkg <-- an mld page created by the driver but compiled like any other pages.
├── Mymodule
├── index.mld <-- Has a reference to 'mydir/index.mld'
├── myimage.gif
├── page.mld
└── mydir/
    ├── index.mld <-- Has a reference to 'index.mld' in the parent directory
    └── myimage.gif
```

Reply:
> Defining what happens in this case does not seem worse than having undefined references exist.

Well it goes without saying that if you use the reference it warns you for the fact it doesn't exist.

> Yes I think we want that!

Actually I'm not sure. I think it would be clearer to have explicit parent notation, that is the moral equivalent of `..` which we all know perfectly is sadly missing from `ocaml`'s module system :-) 

> Maybe it's reasonable to say that we never want to reference a parent page by its name as it was said above ?
Though, the convention should tell how it should be renamed.

We definitively need that if we want to enable packages to escape their little island to link into other ones (in fact AFAIR that worked at some point in `odoc` on the `--pkg` option path).

>     pkg <-- an mld page created by the driver but compiled like any other pages.
>     […]
>     └── mydir/
>         ├── index.mld <-- Has a reference to 'index.mld' in the parent directory

But what do you put on this `pkg` page ? how does it materialize ? 

Basically I think the idea that the parent-child relationship occurs through concrete content (i.e. a .mld page) is not necessarily a good one and we should let it go.

I think we rather want to think in terms of nested bundles: understood as a directory with `.mld` files, assets and possibly module files. The name of the directory defines the reference path/url prefix and the default page is `index.mld` so a hierarchy like: 

```
pkg/ 
├── Mymodule
├── index.mld <-- Has a reference to 'mydir/index.mld'
├── myimage.gif
├── page.mld
└── mydir/
    ├── index.mld <-- Has a reference to 'index.mld' in the parent directory
    └── myimage.gif
```
Generates basically:
```
pkg/ 
├── Mymodule.html
├── index.html
├── myimage.gif
├── page.html
└── mydir/
    ├── index.html
    └── myimage.gif
```

To which you can refer to, as far as the pages are concerned and according to where you are as:

1. In another package.
```
{!page-pkg.index}       # pkg/index.mld
{!page-pkg.page}        # pkg/page.mld
{!page-pkg.mydir.index} # pkg/mydir/index.mld
```
2. When you are in the `pkg` level
```
{!page.index}           # pkg/index.mld
{!page.page}            # pkg/page.mld
{!page.mydir.index}.    # pkg/mydir/index.mld
```
3. When ou are in the `mydir` level
```
{!page-^index}          # pkg/index.mld
{!page-^page}           # pkg/page.mld
{!page-index}           # pkg/mydir/index.mld
```

That way we can actually already have structured documents now. The convention for a package `pkg` is simply as follows:

1. A "directory" `pkg` is created 
2. The content of `odoc-pages` is rerooted to this directory
3. If there is no `index.mld` at that point, generate one for the package in `pkg`
4. The modules are added to this directory.

Directory is methaphorical here this can happen through `--parent` options.

This is a system that is extremely easy to use for end users as it naturally maps on the file system, no crazy renamings going on behind the scene or arbitrary attachements. The position in the file system mostly defines how your reference looks like. The only thing the user needs to be made aware of is the `pkg` directory creation – and only if it needs to make references into other packages.

Reply:
> 1. In another package.
> 
> ```
> {!page-pkg.index}       # pkg/index.mld
> {!page-pkg.page}        # pkg/page.mld
> {!page-pkg.mydir.index} # pkg/mydir/index.mld
> ```

Unless you have a `pkg` hierarchy in your bundle in which case you need to write:

```
{!page-^pkg.index}       # pkg/index.mld
{!page-^pkg.page}        # pkg/page.mld
{!page-^pkg.mydir.index} # pkg/mydir/index.mld
```